worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,131 @@
+  private val detectIpAddressFailover = props.getProperty("org.quartz.dataSource." + dataSource + ".ipFailover") == "true"
+  // Removes "jdbc:mysql://" prefix and ":{port}..." suffix
+  private val dataSourceHostname = if (originalJdbcURL != null) originalJdbcURL.replace("jdbc:mysql://", "").split(":")(0) else null
+
+  // Needs to be mutable so that the pool can be exchanged during a failover AND kept in memory so that connections are long-lived
+  private var pool: Pool = new Pool(getIP)
+  private val poolLock = new Object()
+
+  def createNewConnectionProvider(finalJdbcURL: String): Option[HikariCpPoolingConnectionProvider] = {
+    if(dataSource != null) {
+      Some(new HikariCpPoolingConnectionProvider(
+        props.getProperty("org.quartz.dataSource." + dataSource + ".driver"),
+        finalJdbcURL,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".user"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".password"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".maxConnections").toInt,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".validationQuery")
+      ))
+    } else {
+      logger.info("No job store found in config to get connections")
+      None
+    }
+  }
+
+  /**
+   * HikariCP connection pools don't automatically close when IP addresses for a hostname change. This function returns True, iff at
+   * least one of the following conditions is met:
+   *   - IP addresses have changed for the CNAME record used for DNS lookup
+   *   - causeFailoverEveryConnection is set to "true", which is used for testing failover functionality
+   *
+   * @param pool
+   *   the connection pool currently being used
+   * @param dnsIP
+   *   the IP returned when performing a DNS lookup
+   * @return
+   */
+  private def hasIpAddressChanged(pool: Pool, dnsIP: String): Boolean = {
+    causeFailoverEveryConnection == true || pool.ip != dnsIP
   }
 
-  def getConnection = provider.get.getConnection
+  def retryGettingIp[T](n: Int)(fn: => T): T = {
+    try {
+      fn
+    } catch {
+      // Failed to resolve it from JVM
+      case e if n > 1 => retryGettingIp(n - 1)(fn)
+    }
+  }
+
+  val numRetries = 10
+  def getIP: String = {
+    retryGettingIp(numRetries) {
+      // Get the ip address of the hostname. The result is cached in the JVM
+      java.net.InetAddress.getByName(dataSourceHostname).getHostAddress
+    }
+  }
+
+  def getConnection = {
+    if (detectIpAddressFailover && originalJdbcURL != null) {
+      // If the IP has changed, then we know a failover has occurred, and we need to create a new hikari config
+      val newIP: String = getIP
+      if (hasIpAddressChanged(pool, newIP)) {
+        // A failover has occurred, so we lock the pool and swap it out with a new hikari config
+        val (poolToUse, optionalOldPool) = poolLock synchronized {
```

Comment:
Digging into the source code more, when we call `softEvictConnections()`, all pool entries are marked as evicted, which means they will get removed from the connection bag and replaced with entirely new pool entires. This is important because URL resolution to get the IP happens when a new pool entry is created as shown here: https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/util/DriverDataSource.java#L62

So future connections will get the right IP address when doing a lookup after `softEvictConnections()`. We just need to make sure the jdbc url in the config is the hostname rather than an IP lookup.

I'm wonder if this IP-failover check might just be better suited as a change in the HikariCP repository as an optional configuration. Not an urgent change to make there, but just an idea

Also, I like the idea of using a job to check the ip every so often. I'll come up with something for that

Reply:
Any preference on the type of background job process? We can use quartz job scheduling framework for this, or we can use scala's inherent concurrency libraries

Reply:
Or honestly, a caching system might just be simpler than having a background process that checks if the IP address has changed